File: USACO.GUIDE/SILVER/Graph Traversal/USACO_644.cpp
#pragma GCC target ("avx2")
#pragma GCC optimization ("O3")
#pragma GCC optimization ("unroll-loops")

#include <bits/stdc++.h>

using namespace std;

typedef int64_t ll;

bool FC(unordered_map<ll, vector<ll>> &adj, vector<ll> &closed, ll n)
{   
    vector<ll> visited;
    for (int i = 0; i < n+1; i++)
    {
        visited.push_back(0);
    }
    // dfs
    vector<ll> stack;
    ll start = 1;
    for(start=1; start < n+1; start++){
        if(!closed[start]){
            break;
        }
    }
    stack.push_back(start);
    visited[start] = 1;
    
    while (stack.size())
    {   
        ll node = stack.back();
        stack.pop_back();
        for (int i = 0; i < adj[node].size(); i++)
        {
            if (!visited[adj[node][i]] && !closed[adj[node][i]])
            {
                stack.push_back(adj[node][i]);
                visited[adj[node][i]] = 1;
            }
        }
    }
    for (int i = 1; i < n+1; i++)
    {
        if (visited[i] == 0 && closed[i] == 0)
        {
            return false;
        }
    }
    return true;
}

int main()
{
    freopen("closing.in", "r", stdin);
    freopen("closing.out", "w", stdout);

    ll n, m;
    vector<ll> closed;
    unordered_map<ll, vector<ll>> adj;
    scanf("%ld %ld", &n, &m);
    for (int i = 0; i < n+1; i++)
    {
        closed.push_back(0);
    }

    for (int i = 0; i < m; i++)
    {
        ll a, b;
        scanf("%ld %ld", &a, &b);
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    if (FC(adj, closed, n))
    {
        printf("YES\n");
    }
    else
    {
        printf("NO\n");
    }
    for (int i = 1; i < n; i++)
    {
        ll a;
        scanf("%ld", &a);
        // adj.erase(a);
        closed[a] = 1;
        if (FC(adj, closed, n))
        {
            printf("YES\n");
        }
        else
        {
            printf("NO\n");
        }
    }
    return 0;
}

File: USACO.GUIDE/SILVER/Graph Traversal/USACO_644.py
input = open('closing.in', 'r').readline
fout = open('closing.out', 'w')


n, m = map(int, input().split())
adj = [[] for _ in range(n+1)]
closed = [0]*(n+1)

def FC(adj, closed, n):
    visited = [0]*(n+1)
    stack = []
    node = 1
    for i in range(1, n+1):
        if not closed[i]:
            node = i
            break
    # DFS
    stack.append(node)
    visited[node] = 1
    while stack:
        v = stack.pop()
        for child in adj[v]:
            if not visited[child] and not closed[child]:
                stack.append(child)
                visited[child] = 1

    for i in range(1, n+1):
        if visited[i] == 0 and closed[i] == 0:
            return 0
    return 1


for _ in range(m):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)

pops = []
for i in range(n):
    pops.append(int(input()))




for v in pops:
    if FC(adj, closed, n):
        print("YES", file=fout)
    else:
        print("NO", file=fout)
    closed[v] = 1
